### PR TITLE
fix(create-injection-token): allow provideFn to accept factory

### LINKED ISSUE #99: [RFC] createInjectionToken generic type and optional factory
I'd like to provide some other suggestions.

How about adding support for the **Generic type and marking the factory as optional**?

So basically this:
`export const TOKEN = new InjectionToken<MyType>();`

becomes:
`export const [injectFn, provideFn] = createInjectionToken<MyType>();`

There's a problem though, I can't make it `multi` without providing the factory 🤷🏻‍♂️

Also, considering the code above, would this work?
`provideFn(() => {
  const foo = inject(FOO);
  if (foo) return 'something'
  return 'something else'
})`

Instead of a simple value, to provide a factory!

This is how I manage it now:
image

### PR DESCRIPTION
**Pull request by nartc**
Closes #99

No docs yet because my brain is not firing on all cylinders this morning (Insomnia struck last night so I've been up since 2AM). I'll add docs later or @wanoo21 can put some docs 😋. 

cc @wanoo21 no docs so you might want to take a look at the new tests in this PR for usage.

For "optional factory" case, I had to add a new function `createNoopInjectionToken()` instead. Typing is tricky with the optional factory.

```ts
// "multi" token; unfortunately until I can acquire better typescript skill, we have to duplicate "true" here
const [injectFn, provideFn] = createNoopInjectionToken<number, true>('description', { multi: true });

injectFn(); // number[]
provideFn(1); // accepts number
provideFn(() => 1); // accepts a factory returning a number

// normal token
const [injectFn, provideFn] = createNoopInjectionToken<number>('description');
injectFn(); // number;
```

### CODE REVIEW

**Comment by wanoo21** [comment]
So fast, thanks @nartc, I'll check the source code and test examples this weekend, and I sure can help with docs!

You should take a break, hahaha!

**Review by eneajaho — APPROVED** [review]

**Comment by nartc** [comment]
Waiting on @wanoo21 confirmation before merging.

**Comment by wanoo21** [comment]
I think we're good to move. Let's merge it and I'll update the docs before release.

Should I create a new issue for documentation?

**Pull request merged**
